lib/setMode.js

```diff
@@ -1,21 +1,160 @@
+     */
+    
+    //Send color data for left side
+    sendData(keyboard, 1, leftDef.primary.color, leftDef.primary.level, 0);
+    sendData(keyboard, 2, leftDef.secondary.color, leftDef.secondary.level, 0);
+    sendData(keyboard, 3, period, period, period);
 
-	var commit = new Buffer(8);
-	commit[0] = 1;
-	commit[1] = 2;
-	commit[2] = 65; // commit
-	commit[3] = constants.modes[mode]; // set hardware mode
-	commit[4] = 0;
-	commit[5] = 0;
-	commit[6] = 0;
-	commit[7] = 236; // EOR
-
-	keyboard.sendFeatureReport(commit);
-	keyboard.currentMode = mode;
+    //Send color data for middle
+    sendData(keyboard, 4, middleDef.primary.color, middleDef.primary.level, 0);
+    sendData(keyboard, 5, middleDef.secondary.color, middleDef.secondary.level, 0);
+    sendData(keyboard, 6, period, period, period);
+
+    //Send color data for right side
+    sendData(keyboard, 7, rightDef.primary.color, rightDef.primary.level, 0);
+    sendData(keyboard, 8, rightDef.secondary.color, rightDef.secondary.level, 0);
+    sendData(keyboard, 9, period, period, period);
+  }
+  
+  //Initalize the mode
+  commit(keyboard, constants.modes[mode]);
+  
+  //Save the mode for reference
+  keyboard.currentMode = mode;
+  
 	return keyboard;
 };
 
+function getPrimaryAndSecondarySettingsForRegion(data, region) {
+  if (typeof data === 'string') {
+    if(typeof constants.colors[data] === 'undefined') throw 'invalid color for ' + region + ' region';
+    regionDef = {primary: {color: constants.colors[data], level: 0}, secondary: {color: constants.colors['off'], level: 0}};
+  } else if (typeof data === 'number') {
+    if(data < 0 || data >= constants.colors.length) throw 'invalid color for ' + region + ' region';
+    regionDef = {primary: {color: data, level: 0}, secondary: {color: constants.colors['off'], level: 0}};
+  } else if (typeof data === 'object') {
+    if (data.primary) {
+      regionDef = {primary: getColorAndIntensityLevel(data.primary, region, true)};
+    } else {
+      regionDef = {primary: getColorAndIntensityLevel(data, region, true)};
+    }
+    if (data.secondary) {
+      regionDef.secondary = getColorAndIntensityLevel(data.secondary, region);
+      //Default to off if both settings are unset
+      if (regionDef.secondary.color < 0 && regionDef.secondary.level < 0) regionDef.secondary = {color: constants.colors['off'], level: 0};
+      //Default to primary settings if either are unset
+      if (regionDef.secondary.color < 0) regionDef.secondary.color = regionDef.primary.color;
+      if (regionDef.secondary.level < 0) regionDef.secondary.level = regionDef.primary.level;
+    } else {
+      //Default secondary to off if unset
+      regionDef.secondary = {color: constants.colors['off'], level: 0};
+    }
+  } else 'invalid color for ' + region + ' region';
```

Comment:
This does not return or respond with anything. This should throw an error as it will continue to line 80 and return an undefined variable.

`regionDef` should be defined at the top of the function as null.